File: src/main/java/org/goafabric/personservice/persistence/PersonRepositoryPanache.java
package org.goafabric.personservice.persistence;

import io.quarkus.hibernate.orm.panache.PanacheRepositoryBase;
import io.quarkus.panache.common.Page;
import jakarta.enterprise.context.ApplicationScoped;
import org.goafabric.personservice.controller.dto.PersonSearch;
import org.goafabric.personservice.persistence.entity.PersonEo;

import java.util.HashMap;
import java.util.List;

@ApplicationScoped
public class PersonRepositoryPanache implements PanacheRepositoryBase<PersonEo, String> {

    //todo missing filters with organizationid

    public List<PersonEo> findByStreet(String street, Page page) {
        return find("address.street", street).list(); //page(page).list();
    }


    public List<PersonEo> find(PersonSearch search, Page page) {
        var query = new StringBuilder();
        var params = new HashMap<String, Object>();

        if (search.getFirstName() != null) {
            query.append("firstName = :firstName");
            params.put("firstName", search.getFirstName());
        }

        if (search.getLastName() != null) {
            if (!query.isEmpty()) query.append(" and ");
            query.append("lastName = :lastName");
            params.put("lastName", search.getLastName());
        }

        return find(query.toString(), params).list(); //.page(page).list();
    }


    public PersonEo save(PersonEo person) {
        persist(person);
        return person;
    }

}

File: src/main/kotlin/org/goafabric/personservice/persistence/PersonRepositoryPanache.kt
package org.goafabric.personservice.persistence

import io.quarkus.hibernate.panache.PanacheRepository
import jakarta.data.page.Page
import jakarta.data.page.PageRequest
import jakarta.data.repository.OrderBy
import org.goafabric.personservice.persistence.entity.PersonEo
import org.hibernate.annotations.processing.Find
import org.hibernate.annotations.processing.HQL

interface PersonRepositoryPanache : PanacheRepository.Managed<PersonEo, String> {
    @Find
    @OrderBy("lastName")
    fun findByLastNameAndOrganizationId(
        lastName: String,
        organizationId: String,
        pageable: PageRequest
    ): Page<PersonEo>

    @HQL(
        ("SELECT p FROM PersonEo p WHERE p.organizationId = :organizationId " +
                "AND (:firstName IS NULL OR p.firstName = :firstName) " +
                "AND (:lastName IS NULL OR p.lastName = :lastName)")
    )
    fun search(
        firstName: String?,
        lastName: String?,
        organizationId: String,
        pageable: PageRequest
    ): Page<PersonEo>

    fun save(personEo: PersonEo): PersonEo {
        persist(personEo)
        return personEo
    }
}